Enhancing Filter Functions

KyuaKwon opened the pull request:
- I shorten the getMentorList function and applied a Promise process.
- User.model is edited. Career is no more an array.

I will enhance over and over. Counting function also.
Then I'll do the DB jobs.

( If this PR is merged, the DB should be cleared )

ethkim commented on src/server/controllers/match.controller.js:
```diff
@@ -21,42 +21,100 @@ export const MATCH_STATUS = {
-  let careerFilteredIdList = [];
-  let filteredList = [];
-  let filteredIdList = [];
+  let exceptList = [];
```
exceptionList?

ethkim commented on src/server/controllers/match.controller.js:
```diff
@@ -80,37 +138,52 @@ function arrayContainsElement(arr, val) {
 
 function checkCareerFilter(userInfo, filter) {
```
What does this function do? "check" doesn't tell me what it does - it seems like it's just checking for preconditions for the filter object, but not sure why these are required. Either change the method name or comment on what this function does?

ethkim commented on src/server/controllers/match.controller.js:
```diff
@@ -80,37 +138,52 @@ function arrayContainsElement(arr, val) {
-
-  if (userBool.area && userBool.role && userBool.years && userBool.background) return true;
-  else return false;
+  else return ((filter.area === 'All' || userInfo.area === filter.area)
```
Ouch. This return statement is painfully hard to read - and prone to future errors when other people change. Could we clean up the conditions somehow?

ethkim replied:
Potentially just splitting off the conditions to do the negative:

else if (filter.area != 'All' && userInfo.area != filter.area) return false;
else if (filter.role != 'All' && userInfo.role != filter.role) return false;
else if ...

ethkim replied:
BTW, why do we have this "All" business? Shouldn't we just not have any filter if the filter value is "All"? It seems like a redundant check. Every filter should just be null by default (which implies All)

KyuaKwon replied:
Agree with splitting the conditions.
But about comparing 'All', cannot sure yet.
The conditions(area, role, years, educational_background) are independent so they should be checked individually. Even when 'area' is "All", 'years' needs to be checked.

ethkim commented on src/server/models/users.model.js:
```diff
@@ -70,7 +70,12 @@ let userSchema = new Schema({
   work: [experienceSchema],
-  career: [careerSchema],
+  career: {
+    area: String,
```
If we only support Engineering and Design for now, it'd be good to make these as enum rather than string (same goes for role below). Doesn't need to happen (probably can't..) in this PR, but either leave a TODO or keep that in mind :)

KyuaKwon replied:
We(Dabin, front-end engineer who builds filter part and I) were talked about this before but couldn't reached to cool conclusion. Do you think giving specific number to content is appropriate?

ethkim replied:
I think it makes sense to keep these as specific enum type (together with meta data to represent the string value). That'll save some "string comparisons" later when filtering etc. 
But yeah, totally optional, we could tackle it later.

ethkim commented:
lgtm

KyuaKwon commented:
Few changes again!
- Fixed up some grammar flaws
- Added a TODO to compare the filters with enum
- Change the user's career as an object, not an array
- Updated expired Facebook tokens

Thanks!
